Added some missing C99 float operations

mmottl opened the pull request:
The `Pervasives`-library was missing some more float operations that are part of `<math.h>` in the C99 standard library.  For example, it already implemented hyperbolic functions like `tanh`, but it was missing inverses like `atanh`.

This pull request implements the following functions:

- `exp2`, `log2`: the C99 standard functions should implement them more efficiently and accurately with some bit manipulations.
- `cbrt`: the cubic root
- `acosh`, `asinh`, `atanh`: the missing inverse hyperbolic functions
- `trunc` and `round`
- `erf` and `erfc`: the error function is quite basic but notoriously difficult to implement with good numeric accuracy and high speed

Since there are already some C99-functions in `Pervasives`, it should be fine to add more.  They all come with alternative implementations in the case that a C-compiler does not support C99 (unlikely these days).  All of the above functions have reasonably fast and accurate alternative implementations except for the error function.  The fast and simple algorithm should suffice for users on these rare platforms that don't lend themselves to numeric programming anyway.

The configure script has been updated to check for the availability of these functions.  Only the changelog needs to be updated once it is determined that these functions can be merged.  The changes have been tested in a fresh compiler build and perform exactly as required when C99 was detected.  I have manually checked the rounding errors of the alternative approximations for a range of values, too.

gasche commented:
Wouldn't it make sense to have the tests you performed in the testsuite?

Re. manual implementations, cc: @jhjourdan who may have an opinion on that.

mmottl commented:
@gasche If C99 is detected, the tested wrappers behave exactly as they should, which is easy to test.  Since the wrappers are trivial and will quite surely never change, there seems to be little use for running test suites on them.  There aren't any tests for the already existing C99 wrappers either.

What concerns the alternative implementations, it's not quite clear how these should be tested.  You'd have to force a build with a presumably hacked configure script to avoid C99 detection.  Then you'd have to compare these functions to some reference implementation given a somewhat arbitrary quality criterion.  C99 implementations may differ between platforms.  Verifying algorithms for which only approximate answers exist is not easy...

gasche commented:
You could for example expose `caml_foo_approx` primitives for the approximated version (maybe checking that gcc and clang inline the code as expected in the approximated-or-not version) and bind them as `external` from the testsuite. Then whatever manual test you report that you did on your side could go in the testsuite, and serve to maybe catch systems with bad implementations of the base functions (when our users on exotic machines run the testsuite as part of a porting effort)?

mmottl commented:
I guess it would be overkill to design the OCaml test suite such that it can be used to detect bad library implementations on exotic platforms.  In any case, the test suite would be considerably more complicated than the implementation itself.  I don't think it would be worth it.

Octachron commented on byterun/floats.c:
```diff
@@ -530,6 +650,49 @@ CAMLprim value caml_log1p_float(value f)
+  int sign = (x >= 0) ? 1 : -1;
+  x = fabs(x);
+  t = 1.0 / (1.0 + p * x);
+  y = 1.0 - (((((a5 *t  + a4) * t) + a3) * t + a2) * t + a1) * t * exp(-x * x);
```
Does that line not mean that `simple_erf` is using an asymptotic extension for erfc then using 
erf = 1 - erfc for recovering erf, and then implements `simple_erfc` as `1 - simple_erf`?
Would it not be better to define `simple_erfc` directly here and then implement `simple_erf` as `1 - simple_erfc`?
Or is this round-trip and the associated loss of numerical precision somehow beneficial in this case?

mmottl replied:
The code actually returns `sign(x) * y`.  I think your suggestion would only hold for positive `x`.

Octachron replied:
You are right. I was only looking at the `x>0` branch, where you have erfc(x) = 1 - (1- erfc x) leading to catastrophic cancellation and erfc x = 0 for x ≥ ~6.

mmottl replied:
Numerical precision (e.g. cancellation issues) seems to be of little concern here given that the simple implementation has atrocious numerical accuracy, i.e. is simply a poor approximation.  The latter issue is the far more dominant source of errors.

Octachron replied:
I agree that numerical precision is of little concern here.
Nevertheless, depending on the application, wrongly returning a 0 value for ordinary values of x is not a poor approximation but an atrocious one: for instance, in a probabilistic setting, wrongly returning 0 is as much a correctness issue than returning +∞.

mmottl replied:
Sometimes `0` is just simply the closest answer even a good approximation can come up with.  The C99 implementation on my platform returns `1` or `-1`, too, whenever  `|x| >= ~6`.  The implemented approximation above was published in one of the most widely used reference books (Abramowitz and Stegun: Handbook of Mathematical Functions).  More modern approximations are more accurate, but typically a hell lot more complicated.  Note, too, that some global approximations, which have a guaranteed maximum error in the whole range, may perform much more poorly in ranges that actually matter to people.  I really don't think we need to go overboard here.  To me this is good enough for a fallback solution.

jhjourdan replied:
I think that the code should at least contain a precise reference in comment to this book, if this is where you took it from.

But anyway, I have a larger concern here: it seems that erf(0) is not 0, which is quite annoying, since this means that there is a discontinuity near 0 (btw, is it even positive ? Otherwise that would mean that it is not even increasing). Even if many programs  do not depend on the fact that an approximation is good, many more actually depend on the continuity.

Why focusing on simplicity of the implementation rather than accuracy? If the user does not want accuracy, she can use such a simple implementation herself and including in stdlib is pointless.

jhjourdan replied:
One additional note: if you keep this method, then you could consider using floats instead of doubles. They are much faster and the precision loss is maybe negligible.

mmottl replied:
@jhjourdan On modern CPUs doubles will usually perform better than floats.  Unless memory size or bandwidth is an issue or if there are vectorization advantages with floats, people should use doubles all the time.

I can add a reference if we decide to go ahead with the current implementation, which is based on the formula in the book but implemented using Horner's method, which is better for evaluating polynomials.  I think John Cook is the original author of the reference code I found.

I agree that there is a small discontinuity at zero, but the jump is smaller than one billionth.  Given that we are dealing with finite representations anyway, a small degree of discontinuity is expected.

That said, of course, better approximations exist, especially around zero, where the above approximation is actually least accurate.  But then you'd have to convince the OCaml team of accepting much more complicated code.

jhjourdan replied:
>  On modern CPUs doubles will usually perform better than floats. Unless memory size or bandwidth is an issue or if there are vectorization advantages with floats, people should use doubles all the time.

**Better**? Do you have numbers for this? Anyway, in your case, your are using exp, which is **surely much slower** than expf. And, I presume, this is the bottleneck here.

> That said, of course, better approximations exist, especially around zero, where the above approximation is actually least accurate. But then you'd have to convince the OCaml team of accepting much more complicated code.

In practice what is the precision of this method?

FWIW, at least, I am convinced.

mmottl replied:
I don't know what platform you are using, but I haven't run into practical examples yet where floats significantly outperformed doubles on modern CPUs.  They usually perform the same, but I have seen cases where doubles performed better.  I've just tried it out, and there is no appreciable difference in performance between `exp` and `expf` on my platform.

There are surely cases where floats might be a better choice, because you can fit more of them into the cache or might benefit from better vectorization, but I haven't run into anything like that myself.  There are numerous online discussions about this topic and the general advice is to use doubles unless you have very good reasons not to.  They are almost always at least as fast, sometimes faster, and obviously offer more precision.

jhjourdan replied:
> I've just tried it out, and there is no appreciable difference in performance between exp and expf on my platform.

Well, on my linux machine (glibc 2.19) with a Core i3-2120 @ 3.3GHz, I did the following micro-bench:

```
int main () {
  double sum = 0;
  for(double i = 0; i < 10; i+=1E-7) {
    sum += exp(i);
  }
  printf("%f", sum);
  return 0;
}
```

With `exp`: 2.338s, with `expf`: 1.325s. That's fairly appreciable.

Anyway, I am not saying that this is what we should do. I am just saying that this might be worth trying, just to see how this affects the  precision/performance trade-off.

mmottl replied:
This benchmark looks suspect.  I'd never loop over floating point numbers that way.  If, for example, you replaced `double` with `float`, the loop won't terminate anymore due to cancellation for `i >= 2`.

Anyway, there is no performance difference on my machine for the same code.  I don't know why there is such a big difference on yours.  It's entirely possible that these two functions are implemented differently on your platform, not just for the type.  Just do some basic arithmetic operations in a loop.  There is no difference on my machine between float and double performance in that case.

jhjourdan replied:
> This benchmark looks suspect. I'd never loop over floating point numbers that way. If, for example, you replaced double with float, the loop won't terminate anymore due to cancellation for i >= 2.

Well, give me another benchmark that measures the same thing. I already observed this fact several times in different contexts. And, btw, I get similar results with my other linux machine (Core i5-4300 @ 1.9 GHz, glibc 2.24).

The reason why it is slower is **not** the difference of performance of FP instructions. The reason is that the approximation algorithm used by libc needs to be more involved to get the extra precision.

> Just do some basic arithmetic operations in a loop. There is no difference on my machine between float and double performance in that case.

This is unrelated to this PR, but in practice, if I just do some basic arithmetic operations in a loop, there are some chances that it will be compiled to SIMD instructions, leading to a larger throughput.

oandrieu commented:
Incidentally, recent MSVC versions have adequate C99 support so we might want to add:
```c
#if defined(_MSC_VER) && _MSC_VER >= 1800
#define HAS_C99_FLOAT_OPS
#endif
```
in `s-nt.h`

jhjourdan commented:
Not mentioning the fact that none of these implementations are correctly rounded (nor C99 ones are, usually), I'd like to -at least- make sure that they do something sensible for extreme/infinite/NaN/null values of their parameters.
It seems, for example, that your implementations of inverse hyperbolic trigonometric functions break for large values of their parameters.

mmottl commented:
It's important to keep in mind that we are merely implementing fallback functions for platforms that should likely not be used for scientific computations anyway.  There is a good reason why C99 introduced these functions, because programmers typically get them wrong.

Our options here range from "raise an exception" to "implement the state of the art".  I don't think anyone here would go to either extreme.  The "state of the art" can be pretty intimidating.  Just look at a typical high quality implementation, e.g.: [Sun's `erf`](http://www.ks.uiuc.edu/Research/namd/doxygen/erf_8C-source.html).  Or check out this page [discussing  `asinh`](http://stackoverflow.com/questions/34535278/most-accurate-way-to-compute-asinhf-from-log1pf/34548547).

I'm perfectly happy to make small, portable improvements to our fallback implementations.  Or, if copyright or potential maintenance is not an issue for the OCaml team, ship a portable high quality implementation.  But do we really need to offer complex state of the art numerical algorithms to handle mathematical corner cases on exotic platforms?  I'd go with the simple solution here.

gasche commented:
I agree that using naive (not state-of-the-art-efficient) implementations should be fine for the purpose you describe. That does not mean that we shouldn't be careful about correctness, though. My point about testing was that if you did certain tests to make sure that these implementations behave "reasonably" (including the tests on corner-case values that Jacques-Henri just mentioned), these tests would profitably go in the testsuite. I think that this is reasonable, not overkill.

mmottl commented:
@oandrieu I don't know about configuration issues on Windows.  Is it really necessary to modify `s-nt.h`?  I would think that the already existing detection code in `configure` for C99 should work there, too.

oandrieu commented:
@mmottl : the `configure` script is exclusively for Unix. Configuration for the Windows ports is hardcoded in the relevant files in `config/`.

Chris00 commented:
On 2016-12-01, Markus Mottl wrote:
>
> Numerical precision (e.g. cancellation issues) seems to be of little 
> concern here given that the simple implementation has atrocious 
> numerical accuracy, i.e. is simply a poor approximation.  The latter 
> issue is the far more dominant source of errors.

Why not use the version in http://openlibm.org/ ?

mmottl commented:
@Chris00 I would personally be fine with adding more sophisticated implementations if copyright and potentially more maintenance is not an issue for the OCaml team.  `openlibm` apparently uses some other existing free implementations anyway, e.g. Sun's `erf` mentioned previously.  Again, this should just be a fallback mechanism.  I'd prefer using available platform-specific C99 implementations.

mmottl commented:
@oandrieu Thanks, I didn't know that `configure` was for Unix only.  I have added your suggested Windows patch to the pull request.

shindere commented:
Markus Mottl (2016/12/01 06:44 -0800):
> @oandrieu I don't know about configuration issues on Windows.  Is it
> really necessary to modify `s-nt.h`?  I would think that the already
> existing detection code in `configure` for C99 should work there, too.

N, configure can't be used under Windows yet.

mmottl commented:
In case anybody wants to see state of the art implementations of missing C99-functions, here is a link to [glibc](https://github.com/bminor/glibc/tree/master/sysdeps/ieee754/dbl-64).  As you can see there, even some of the functions that already exist in OCaml are considerably more complex than the OCaml fallback implementations.

objmagic commented:
@mmottl while you are here, do you think it is reasonable to also add function that sets rounding mode? Frama-C, when doing abstract interpretation on floating point operations, has to use C bindings to control rounding mode. This is a bit inconvenient (See: [c_bindings.c](https://github.com/Frama-C/Frama-C-snapshot/blob/5b1ef6ae1ed11c55f0d2e1abe740fc30e7b2bdac/src/libraries/utils/c_bindings.c#L191))

mmottl commented:
@objmagic There are surely many more missing C99 functions (e.g. `tgamma`, `lgamma`).  I think they should eventually be added, too, but I had focused on the more frequently needed ones.  We don't need to add everything that's missing right away.  Rounding mode support could be added in another pull request.

mshinwell commented:
As far as I know control of the rounding mode is not provided from OCaml due to the desire to expose a common semantics at all times across all ports.  Please start a separate pull request to discuss this further if required.

silene commented:
> It's important to keep in mind that we are merely implementing fallback functions for platforms that should likely not be used for scientific computations anyway. There is a good reason why C99 introduced these functions, because programmers typically get them wrong.

The thing is, your fallback functions might be just as wrong as typical implementations, so by providing them to unsuspecting Ocaml users, you are actually making them a disservice. I think it is much more honest, even if you feel that to be extreme, to throw an exception rather than to give a broken approximation.

More concretely, there are various issues in addition to those already mentioned:

- The floating-point evaluation of `log(x + sqrt(1 + x*x))` is not a suitable approximation of `asinh` for inputs close to zero, for reasonable negative inputs, and for large positive inputs. One of the proper ways to implement a (poor) approximation of `asinh` is to use `log1p(x + x*x / (1 + sqrt(1 + x*x)))` for small inputs, `log(2*x)` for large positive inputs, and `-asinh(-x)` for negative inputs.

- The floating-point evaluation of `0.5 * log((1 + x) / (1 - x))` is not a suitable approximation of `atanh` for inputs close to zero. One of the proper ways to implement a (poor) approximation of `atanh` is to use `0.5 * log1p(2*x / (1 - x))` for small inputs. I believe this formula is also good enough for inputs closer to -1 or 1 but this needs some double-check.

- The floating-point evaluation of `log(x + sqrt(x*x - 1))` is not a suitable approximation of `acosh` for inputs close to one and for large inputs. One of the proper ways to implement a (poor) approximation of `acosh` is to use `log1p((x - 1) + sqrt((x - 1) * (x + 1)))` for inputs close to one and `log(2*x)` for large inputs.

- Your implementation of `round` might fail to be the identity function for large odd integers (e.g. for 2^52 + 1).

mmottl commented:
I completely agree that some of the proposed fallback implementations may be inadequate for demanding numerical applications.  The problem is that the moment we start deviating from simple mathematical identities by branching based on input ranges, etc., then we'll have to provide carefully crafted test suites and/or theoretical proofs of precision guarantees.  Numerical programming at that level is hard and tedious.  Mistakes can easily slip through.

I'd be totally happy if the OCaml team were to accept high quality implementations from, say, the GCC distribution as fallback solutions.  I have full confidence that its developers have already done their homework (proofs, test suites, significant real world use).  No need to reinvent the wheel when you can copy jet engines.

damiendoligez commented:
I'm suspending this PR for the same reason that I suspended #964: to avoid adding more things in the global namespace. Both PRs will be resumed after #1010 is merged.

mmottl commented:
@damiendoligez That's certainly a good idea.  I'd also opt for introducing a `Float`-module as proposed in #964.

alainfrisch commented:
Unsuspending now that #1010 has been merged.  Who will be the first to propose a PR with the new Float module?

pascal-cuoq commented:
I would like to reiterate pretty much what silene said, if perhaps just only a different enough point of view to justify a comment of my own.

Crappy fallbacks are crappy. They make the world worse just by existing, even if they are rarely used. Years ago, I was working on making Frama-C reproducible and correct—for some definition of correct—for floating-point constructs, and that involved writing my own versions of functions that I would have liked ideally to exist in reproducible and correct versions already. It was a waste of time every time to find another implementation of, say, conversion to or from decimal, look at it a while, discover that it introduced several ULPs of error, and move on to the next independent but equally crappy implementation. The fallbacks of Tk and Perl were among the time-wasters. Please do not add OCaml's to the list of bad results in this particular google search.

My suggestion would be to let the first OCaml user who does not have a C99 compiler and is not satisfied with the new functions being implemented as exceptions manifest themselves.

(I think the `round` function in this PR also produces the wrong result for the float immediately before 0.5.)

mmottl commented:
Though it would likely never affect me personally, I'd not be happy about raising exceptions on platforms that don't come with their own high quality implementations.

If most people feel it should be high quality or bust, I'd opt for high quality.  My recommendation at this point would be to adopt the license-compatible, high quality implementations by Sun Microsystems, which have been battle-tested for more than two decades in many projects, including GCC and Python.

alainfrisch commented:
(Related: PR #1354 )

@mmottl : can you provide a pointer to the Sun library you have in mind?  I could not find an obvious candidate after Googling for 2 minutes.

mmottl commented:
@alainfrisch You may want to take a look at [OpenLibm](http://openlibm.org).  It's basically most of the old Sun `libm` library with some modifications to make it more portable.  Here you can take a look at the [function implementations](https://github.com/JuliaMath/openlibm/tree/master/src).  They are obviously extremely well-tested and efficient, but the implementation is certainly not for the faint of heart.  Basically, just accept it works :)

alainfrisch commented:
Ok, thanks.  I've no strong opinion, but my feeling is that it will be difficult to add a dependency to such a project in the core distribution.  This might be a bit too specialized for the general audience of the stdlib and be better dealt with by a dedicated library.  Is another maintainer willing to champion the idea ?  If not, what should we do with this PR?

Chris00 commented on otherlibs/threads/pervasives.ml:
```diff
@@ -117,14 +117,22 @@ external ( /. ) : float -> float -> float = "%divfloat"
 external ( ** ) : float -> float -> float = "caml_power_float" "pow"
   [@@unboxed] [@@noalloc]
 external exp : float -> float = "caml_exp_float" "exp" [@@unboxed] [@@noalloc]
+external exp2 : float -> float = "caml_exp2_float" "caml_exp2"
```
Shouldn't all these functions be in the `Float` module?

mmottl replied:
Yes, of course.  This pull request is really old, older than the `Float` module.

mmottl commented:
@alainfrisch For me personally it's not a big issue, because I can always just use my own bindings for that purpose.  The most important argument to consider is that almost all relevant platforms support C99 and hence already implement these functions in an efficient and numerically stable way.

Ultimately, I see these choices:

1. Don't support these functions.  Most people likely don't need them ever and the others do rarely, but they are fairly widespread in other languages, especially ones frequently used for numerical programming.  It's not exactly a badge of honor to skimp on that functionality if OCaml wants to be seen as attractive for numerical programming.
2. Provide simple fallback solutions that maybe don't satisfy discriminating data scientists, but may be sufficient for casual use.
3. Raise exceptions on platforms that don't fully support C99.
4. Add sophisticated fallback solutions, e.g. using the old Sun implementations or shipping with something like OpenLibm.

My preferred option would be 4) if it doesn't bother the OCaml team to add some more bulk, but I don't feel strongly about that.  I've come around a bit on the next option: maybe raising exceptions as a fallback is the better default, but then it should be possible to build OCaml with a flag that chooses the simple fallback solutions.

Again, we are maybe overthinking this.  Almost all relevant platforms, certainly ones favored for numerical programming, already provide these C99 functions.  We are just deciding on how to handle rare corner cases.

pmetzger commented:
C99 is nearly twenty years old (more than old enough to vote or sign contracts in most jurisdictions) and there are multiple open source implementations.

alainfrisch commented:
Concretely, which of the platforms supported by OCaml would not have the C99 functions?  When were they introduced in MSVC, for instance?

nojb commented:
Roughly VS 2013 for the functions in this PR (https://blogs.msdn.microsoft.com/vcblog/2013/07/19/c99-library-support-in-visual-studio-2013/)

pmetzger commented:
So they've been in VS for five years.

mmottl commented:
Having taken a look at the C99 compatibility status of some major and not so major compilers, I believe it's safe to say that virtually every C compiler with reasonably widespread use supports the above math functions.

alainfrisch commented:
OCaml currently supports Visual Studio back to VS 2005 (according to README.win32.adoc).  I'm personally not shocked if the missing functions are filled with failing stubs (or naive emulations that also dump a warning on stderr), but perhaps @dra27 or someone else would not be happy with that.

pmetzger commented:
@dra27 is the subject matter expert; if he is okay with OCaml only supporting back to VS 2013, then I think that's okay and the README could be updated.

dra27 commented:
The detection is not quite right. VS2013 doesn't (according to the docs) have `exp2`, `log2` or `trunc` which were added in VS2015.

A few thoughts on it:
 - Support for *new* functions should not be held back by old compilers, especially where these functions aren't actually that new!
 - I don't agree with @alainfrisch's suggestion of printing things to stderr (cf. `FAST_CWD` mistake in Cygwin which still haunts the Cygwin years later...). Have an exception or other way of detecting the lack of support
 - The documentation in pervasives.mli should document whether the alternative is potentially inaccurate or slower, as appropriate
 - If the functions don't raise exceptions when the C99 support isn't there, it would be sensible for there to be a way to detect that the support is actually present - i.e. something like `Sys.has_c99_support`
 - There's no supported version of Windows which can't be targeted by VS2015, so I don't think there's as much need for sophisticated fall-back support - the Windows user who must have these functions on an old compiler can use a third party lib, as suggested.

nojb commented:
> The detection is not quite right. VS2013 doesn't (according to the docs) have exp2, log2 or trunc which were added in VS2015.

They are mentioned here as if they **are** present in VS 2013, though: https://blogs.msdn.microsoft.com/vcblog/2013/07/19/c99-library-support-in-visual-studio-2013/; is this wrong?

dra27 commented:
@nojb - yes, indeed - I just checked the CRT. They are present in the headers if not, apparently, in the documentation!

pmetzger commented:
So if it's in VS2013, five years would seem to be enough time for people to upgrade to me at least?

alainfrisch commented:
One issue with failing stubs is that this prevents (AFAIK) marking the externals with "noalloc", and thus degrades the performance.  Can anyone confirm that noalloc is incompatible with raising functions?  (cc @damiendoligez)  If confirmed, one might want to go with either naive emulation or NaN-returning functions.

nojb commented:
Indeed, raising functions cannot be marked `noalloc` (the exception handler is not made available to such functions).

alainfrisch commented:
Ok, so the question is whether the fallback should return NaN or a naive emulation.  I'm leaning towards returning NaN (less likely to go unnoticed).

mmottl commented:
Indeed, raising exceptions is incompatible with `noalloc` so that's maybe not the best solution then.  Another possibility would be to call `abort()`.  This would help users identify the problem more easily than wondering about the origin of NaNs, which might not necessarily indicate an error condition.

dbuenzli commented:
> Indeed, raising exceptions is incompatible with `noalloc` so that's maybe not the best solution then. Another possibility would be to call `abort()`.

To be honest I would largely prefer if I hadn't been able to build the compiler in the first place (same with the `nan` solution) otherwise your program may be running fine for days until it hits a path that uses these functions. The `nan` solution would make things even more obscure.

Why not simply ask for `C99` and let users who can't satisfy this requirement hack around their system to get `openlibm` in their C compilation environment ?

mmottl commented:
Maybe this is a good time to address the more general problem of how the compiler should reflect missing platform features in APIs.  E.g. one could generate APIs like:

```ocaml
val erf : [ `C99_math_unsupported ]
```
instead of:

```ocaml
external erf : float -> float = "caml_erf_float" "erf" [@@unboxed] [@@noalloc]
```

This would allow you to build the compiler as well as supported software just fine on obscure platforms.  But unsupported code would yield precise and easily understood compile time errors.

nojb commented:
I think we try to avoid having signatures depend on configure-time settings as it becomes hard to test.

What about returning NaN **and** printing a runtime warning (cf `Sys.enable_runtime_warnings`)?

mmottl commented:
@nojb Fair enough.  Only returning NaN for unsupported functions seems like a bad idea, because NaN could conceivably be the correct answer to an ill-posed problem.  Maybe using simple fallback functions + emitting a runtime warning on the first use of a potentially inaccurate approximation would be fine, too, then?  The user would have to read the warning anyway to determine what to make of the result.  Maybe they are fine with less than optimal approximations.

alainfrisch commented:
Changing signatures would not really make sense for bytecode anyway (what matters is the availability of the primitives in the runtime -- e.g. ocamlrun -- not on the machine that compiled the bytecode).

mmottl commented:
@alainfrisch Good point about bytecode, the signature modification would obviously be meaningless in that context so that's not a viable solution.

alainfrisch commented:
@dra27 : you were against the idea of emitting a runtime warning.  What would be your suggestion here?

> Why not simply ask for C99 and let users who can't satisfy this requirement hack around their system to get openlibm in their C compilation environment ?

Indeed, it might be time to accept the requirement.  @xavierleroy or @damiendoligez : do you have an opinion on this?

dbuenzli commented:
> hack around their system to get `openlibm` in their C compilation environment ?

Just to get an idea what that means in practice. I once had to integrate this in another setting and basically it was a matter of defining a `math.h` with: 

```
#ifndef MATH_H
#define MATH_H
#include <openlibm.h>
#endif
```
rename `libopenlibm.a` to `libm.a` and tweak the `LIBRARY_PATH` and `C_INCLUDE_PATH` environment variables for the C compiler to lookup these entities first.

The supported platforms of `openlibm` are listed [here](https://github.com/JuliaMath/openlibm#platform-support).

I think we can let package managers handle these quirks rather than cripple the runtime behaviour of our OCaml programs.

alainfrisch commented:
Somehow related: https://caml.inria.fr/mantis/view.php?id=7673 .  This would need to be double-checked with recent versions, but mingw64 can have terrible performance for exp/cos, and probably other C99 math functions.  One conclusion is that some users might need to plug an external math library even if the "official" platform libm provides the function.  It would be interesting to see if what @dbuenzli suggested in the previous note would work under Windows.

dra27 commented:
Some further thoughts on this:

 - I agree that it's completely reasonable to expect a C99 compiler on any non-Windows platform, and also, therefore, for Cygwin.
 - Given that, I think the `configure` tests can be converted to a hard error (but still define `HAS_C99_FLOAT_OPS`)
 - My floating point fu is not brilliant - are the alternate implementations here, in terms of the relevant standards, wrong, or just occasionally inaccurate in a way which, given the age of C99, one might not expect?
 - For _MSC_VER < 1800, the hard error should become a warning and then shim versions be enabled - probably by defining a `HAS_BROKEN_` variant, or some such (see the printf shims already there for floats). The documentation for the ports should also reflect this.

I toyed with having an option where the functions simply weren't present, but it's just too weird to have different editions of the compiler with fundamentally different checksums for standard library modules.

Being able to plug in openlibm is a good feature, and workaround, but I don't think it's necessary to hold up this GPR further by exploring it for these functions. It could definitely be worth revisting in MPR#7673 (I'm messing about with native Windows containers for OCaml at the moment - I'll see how the microbenchmark in that MPR fares).

For what it's worth, the yardstick for MSVC support is not how old the compiler is, it's a function of whether Microsoft still support that compiler, whether Consortium(/Foundation) members retain an interest in that compiler, whether we have core devs willing to do the work and finally whether the workarounds are simple enough to be worth maintaining. At the moment, that function is still returning [true]...

mmottl commented:
I agree that having different APIs depending on platform features has too many downsides so that alternative is not on my list anymore.  This would then also exclude the exception-raising alternative if people want to keep `noalloc` for performance reasons on platforms that support the functions (the vast majority).

How about just failing during configuration by default if these functions are unavailable?  This would only happen on quite obscure platforms anyway.  A configure flag could then be used to override the error, possibly with a choice for the user: return NaN or use the simple fallback implementation.  That's easy to implement, integrate with package managers, and makes it explicit to the user what the implications will be.

It's usually a good idea to wait until people actually complain about a missing feature before implementing it.  If people really want to be able to override functions with e.g. `openlibm`, we can discuss this in a separate pull request in the future.

pmetzger commented:
> How about just failing during configuration by default if these functions are unavailable? 

+1

damiendoligez commented:
Here's my suggestion: let's fail at configure time if the functions are not available, and add an entry at the end of `INSTALL.adoc` to suggest using openlibm in that case. I don't think we should even bother providing fallback implementations.

IMO, returning NaN is just the second-crappiest possible fallback implementation: because NaN is a just a float, it could make the program return incorrect results instead of crashing.

(The crappiest possible fallback would be to return 0, but NaN is close.)

alainfrisch commented:
Ok, there seems to be a strong support for requiring the functions at configure time.  Let's go for it.

Should we then do the same for fma (introduced in #1354) instead of keeping the current fallback implementation?

dra27 commented:
@alainfrisch FWIW, I'm on the point of a GPR to fix Cygwin64 which requires that fma fallback (since newlib's fma is unimplemented!).

dra27 commented:
@damiendoligez - removing the fallbacks changes the profile of supported compilers... do we definitely want that?

mmottl commented:
I'd be fine with @damiendoligez's  suggestion to just fail at configure time as long as `configure` supports alternative `libm` installations.  We can then let OPAM or system package managers handle installing `libm` alternatives (`openlibm` also supports `fma` btw.) and configuring the compiler.  That way the OCaml distribution would remain free from complicated numerical algorithms.

damiendoligez commented:
@dra27 - do you mean dropping support for old versions of MSVC or is there more to it? Do you think these old versions are not old enough to be comfortable dropping them?

dra27 commented:
@damiendoligez - I don’t think compilers which are included in supported SDKs should be dropped (we still test the Windows 7 SDK on AppVeyor). Someone forced to use an old MS compiler (OS support, other binary support, whatever) probably doesn’t care about the C99 math functions, but will care that they can’t build OCaml with it. Given that the emulated functions are written, and shouldn’t change, I’d prefer to leave them in so that there’s a “quick” way to build the compiler still.

damiendoligez commented:
@dra27 What about leaving the (imperfect) fallback functions in the code, but demanding a configure argument to enable them? And documenting the fact that they are imperfect?

dra27 commented:
@damiendoligez - that makes a lot of sense, yes.

@mmottl - there's quite a rebase to do to this, including redoing the `configure` stuff for autoconf. Would you be happy to do this, or would you like one of us to take it from here and rebase it? (I'd be happy to help with the Windows side, regardless)

mmottl committed:
Added some missing C99 float operations

mmottl commented:
I have just rebased this pull request.  The new version is mostly equivalent to the previous patch, but adds the new C99 float operations (if not already present) to `Stdlib` and `Float`.

The imprecise `erf` implementation has been documented.

`configure.ac` has been updated (but not `configure` yet) and has a new flag: `--enable-imprecise-c99-float-ops`.  Its name hopefully indicates clearly enough that something imprecise is going on.  The semantics is as follows: if all required C99 operations are found, they will be used irrespective of the value of this new flag.  If C99 operations are missing and if the user has not explicitly enabled this flag, an error will occur during configuration.  Otherwise _all_ C99 functions will use the imprecise implementation.

This may not seem fine-grained enough for some.  However, chances are that if some C99 float operations are available, then all are.  Almost all platforms should support these C99 operations anyway.  A more fine-grained solution would require many more macro definitions.

xavierleroy commented:
From a distance this looks good to me.  However I'm not sure about what happens with the only major platform known not to support C99 functions, namely Windows/MSVC and Windows/mingw.  Are we supposed to always `configure --enable-imprecise-c99-float-ops` on this platform, otherwise compilation will fail?  Or is the `imprecise-c99-float-op` mode automatically activated on this platform?

mmottl commented:
There have been some changes to files related to Windows support.  In particular, `s-nt.h` is not there anymore.  I've only touched `configure.ac` for platform configuration, which is probably still not used at all on Windows.  I'd expect that even remotely recent Windows environments support all required C99 functions.  It's probably just a matter of making sure that `HAS_C99_FLOAT_OPS` is set somewhere for this platform and that all Windows-specific header files are included.  I unfortunately cannot test this.

xavierleroy commented:
`configure` is now used for the Windows ports, see https://github.com/ocaml/ocaml/blob/trunk/README.win32.adoc .  And I'm skeptical that Microsoft's C run-time system implements all these C99 FP operations.  @dra27, what do you think?

bschommer commented:
Mircosoft's C run-time system implements more of the C99 FP operations since 2013, see [C99 library support in Visual Studio 2013](https://devblogs.microsoft.com/cppblog/c99-library-support-in-visual-studio-2013/).

copy commented on stdlib/stdlib.mli:
```diff
@@ -553,6 +574,66 @@ external tanh : float -> float = "caml_tanh_float" "tanh"
+   rounding direction.  If [x] is an integer, [+0.], [-0.], [nan], or
+   infinite, [x] itself is returned.
+
+   @since 4.08.0 *)
```
Shouldn't these functions only be added to the new `Float` module? This PR adds them to both `Float` and `Stdlib`.

mmottl replied:
I had thought about this and decided to add them to both for now until other people can weigh in.  Obviously, `Stdlib` already contains a large number of float operations.  A lot of code would break if we were to remove all of them from there.  It's not clear to me why only some standard float operations should remain while others should not, but I'm open for other people's take on this.

dbuenzli replied:
I think there's a consensus to no longer add directly to `Stdlib`. All recent floating point operations additions (#1354, #2177, #1794) only went to the `Float` module.

mmottl replied:
I can certainly remove the `Stdlib` functions that were previously in `Float` only (`trunc`, and `round`) and some of the new ones.  But maybe some of the new functions should stay.  E.g. we have `sin`, `cos`, `tan` as well as their inverses `asin`, etc., in `Stdlib`.  We also already have `sinh`, `cosh`, and `tanh` there.   It would seem weirdly inconsistent to not have the hyperbolic inverses (`asinh`, etc.) in `Stdlib`.

Considering the feedback, my suggestion would be to remove all additions to `Stdlib` from this patch except for `asinh`, `acosh`, and `atanh`, which should remain for consistency.

dbuenzli replied:
I'm not the one who has the power on this. But my opinion is that if people want to be consistent they should always use the `Float` module and forget about the  `Stdlib` entry points. Besides these `Stdlib` entry points should be deprecated once the `Float` module hits debian stable, despite the fact that some people want to make us believe that deprecating stuff breaks code (secret: it does not). Then in n more years we might be able to delete them altogheter.

mmottl replied:
I'm generally in favor of eventually moving all float operations to `Float`.  But given that this will likely take a very long time, I don't see the harm in having a consistent set of float operations in `Stdlib` until then.  It's almost certainly not going to make programs any more dependent on deprecated functions than they would be otherwise.  Anyway, I personally don't have strong feelings about this, since I'll switch to using `Float` only in either case.

gasche replied:
Personally I would go with @mmottl's proposal to remove everything from stdlib except the reverse hyperbolic functions. It sounds like a good compromise to me.

mmottl commented:
Ok, I've just updated and squashed the commits into one.  `Stdlib` now only adds the new inverse hyperbolic functions.  This should make sense, because the regular hyperbolic functions are already there, too.  Only `Float` contains all the novel functions.

The only file that might still need updating is `configure`.  Please let me know if this should be part of the patch, or whether you usually update it separately before releases.

gasche commented:
I would update `./configure` as part of the patch.

mmottl commented:
I've updated `configure` by running `autogen`, which is apparently the right way to do that.  There is now also a `Changes` entry for this patch.  After three years it seems we are slowly getting closer to something that people like.

xavierleroy commented:
I ran "CI precheck" on this pull request to test it on many different platforms.

The good news is that it builds and tests fine on all platforms tested.

The not-so-good news is that I still don't understand what is going on with the Windows (MSVC and Mingw) platforms.  `configure` (without `--enable-imprecise-c99-float-ops`) does not test for the presence of the C99 operations.  It undefines `HAS_C99_FLOAT_OPS`, causing the emulations to be used.  This is strange for two reasons:
- The intended behavior, as far as I understand, was for `configure` to fail and ask to be re-run with `configure --enable-imprecise-c99-float-ops`.  But this is not what happens under Windows.  Is it intended?
- As @bschommer wrote, recent enough versions of the MS CRT do provide most (all?) of the C99 FP operations, so we might as well test for  their presence and use them if they are there.

@dra27 @shindere : any opinions?

dra27 commented:
I had a look yesterday - the logic is definitely not right, since AppVeyor tests an old msvc. I didn’t have time to poke it, but it seemed better to use `AC_CHECK_FUNCS`. I’m happy to work on this, given that the combination of autoconf, Windows and old MSCRT will not fill many with joy!

shindere commented:
David Allsopp (2019/11/04 01:33 -0800):
> I had a look yesterday - the logic is definitely not right, since
> AppVeyor tests an old msvc. I didn’t have time to poke it, but it
> seemed better to `AC_CHECK_FUNCS`. I’m happy to work on this, given
> that the combination of autoconf, Windows and old MSCRT will not fill
> many with joy!

Thanks a lot, @dra27. I don't feel really an expert in this, I have to
say.

damiendoligez commented on configure.ac:
```diff
@@ -1052,9 +1057,37 @@ AS_CASE([$host],
   AS_IF([$has_c99_float_ops],
     [AC_CHECK_FUNC([copysign], [AC_DEFINE([HAS_C99_FLOAT_OPS])])])])
 
+AS_IF([$has_c99_float_ops],
```
This doesn't look right. When `copysign` is the only missing function, `$has_c99_float_ops` is true, but `HAS_C99_FLOAT_OPS` is left undefined.

mmottl replied:
Thanks, good catch, the old `AS_IF` for `copysign` needed updating, too.

mmottl commented:
I've fixed the last problem reported by @damiendoligez and rebased the PR to the latest trunk.

damiendoligez commented:
Note that I don't feel qualified to approve this PR without @dra27's review.

gasche commented:
Let's get this over, shall we? @dra27, I believe this needs your re-review and approval (... when you have time, of course!), and then this should be good to merge.

gasche commented:
(I will close and reopen now to force a CI re-run.)

gasche commented:
The testsuite is failing, but this seems to be an innocuous failure due to the location of stdlib.ml functions changing. This should be easy to rebase. I'm fine with doing myself, provided we have an approval, for example from @dra27.

dra27 commented:
It needs https://github.com/ocaml/ocaml/pull/944#issuecomment-549276582 addressed

silene commented:
Note that all the documentation comments state `@since 4.10.0`.

mmottl commented:
I've just fixed the location of the changelog entry and updated the `@since 4.10.0` comments to `@since 4.13.0`.

dra27 commented:
Thanks, @mmottl - I have configure and build system branches coming out of my ears at the moment, and will hopefully sort out the autoconf parts of this either later today or this week! Do you mind my forcing rebasing/pushing to this PR directly?

mmottl commented:
@dra27 Thanks, please feel free to make any necessary changes to the PR.

dra27 committed:
Fix pr2195 reference files

dra27 committed:
Simplify with AC_CHECK_FUNCS

dra27 committed:
Simplify configure test

dra27 committed:
Update float.template.mli

dra27 committed:
Automatically enable shims on VS 2012 and earlier

dra27 committed:
Fix ocaml_cv_cc_vendor on Windows

Trailing \r in the value...

dra27 committed:
Use Caml_inline

dra27 committed:
Refactor lib-float test

Always run all the tests (cf. fma test) and return more failure
information in the result.

dra27 committed:
Workaround broken round on mingw-w64

dra27 committed:
Include classify_float on error in fma

dra27 committed:
Generalise the fma test in configure

dra27 committed:
Add autoconf test for runnable host exes

When building for Windows, configuring with --build and --host enables
cross-compilation mode but configuring with just --host doesn't (since
the resulting executables can be run on the build machine). Adds an
additional check to see whether executables from the C compiler can be
run _regardless_ of autoconf's cross-compilation mode.

dra27 commented:
Rebased with minor fixes to get the Unix CI side to pass. I've opened https://github.com/mmottl/ocaml/pull/1 to discuss the Windows changes.

dra27 commented:
Right, Windows parts pushed. They:
- Include some noisy, but essentially minor, updates to the lib-float and fma tests to include more information when they fail (especially for lib-float, the entire test is now run).
- There are problems with `fma` and `round` on Windows (with tracked bugs for both upstream). I don't wish mingw-w64 to require a new configuration switch in order to build without error, especially as that would result in the user carrying on using emulation even after the bug is fixed. I have therefore introduced cross-compilation-compatible configure-tests for `fma` and `round` specifically.

On all Unix platforms (apart from Cygwin), absent C99 float functions or detectable broken `round`/`fma` are **hard errors** unless `--enable-imprecise-c99-float-ops` is given.

`--enable-imprecise-c99-float-ops` is the default for:
- VS 2012 and earlier (no C99 float ops at all)
- VS 2017 and earlier for `fma` only (which have bugs fixed in VS2019)
- x64 mingw-w64 for `round` and `fma` only (which have bugs)
- Cygwin64 for `fma` only (which is fundamentally broken)

There is some follow-up work, because `fma` is not correct on OpenBSD, but that can be dealt with separately to this long-overdue-to-be-merged PR.

This PR is going through [precheck again](https://ci.inria.fr/ocaml/job/precheck/598/), but I have checked the `configure` output of the previous run 597 thoroughly.

Octachron commented:
I think that the documentation of the new functions need to be updated to reflect the fact that `enable-imprecise-c99-float-ops` is always enabled on Windows except with very recent of VS studio due to portability issue.

dra27 commented:
Just to clarify: the wholesale emulation these functions only happens on VS2012 and older. It's just `fma` and `round` which are affected (both of which were already present as they were added in 4.08, but always in emulated mode on Windows). For all the others, they're correct on mingw-w64 and for VS2013+ (which is quite old).

Does this look sufficient for the documentation for those two functions?

```diff
diff --git a/stdlib/float.mli b/stdlib/float.mli
index 3a2f51d411..2c8583874c 100644
--- a/stdlib/float.mli
+++ b/stdlib/float.mli
@@ -72,7 +72,12 @@ external fma : float -> float -> float -> float =
    instructions (providing full IEEE compliance) or a software
    emulation.  Note: since software emulation of the fma is costly,
    make sure that you are using hardware fma support if performance
-   matters.  @since 4.08.0 *)
+   matters.
+
+   On 64-bit Cygwin, 64-bit mingw-w64 and MSVC 2017 and earlier, this function
+   may be emulated owing to known bugs on limitations on these platforms.
+
+   @since 4.08.0 *)

 external rem : float -> float -> float = "caml_fmod_float" "fmod"
 [@@unboxed] [@@noalloc]
@@ -346,6 +351,9 @@ external round : float -> float = "caml_round_float" "caml_round"
    rounding direction.  If [x] is an integer, [+0.], [-0.], [nan], or
    infinite, [x] itself is returned.

+   On 64-bit mingw-w64, this function may be emulated owing to a bug in the CRT
+   on this platform.
+
    @since 4.08.0 *)

 external ceil : float -> float = "caml_ceil_float" "ceil"
```

Octachron commented:
Indeed, I missed that point; it seems fine to not document the limitation of VS2012 in our documentation.
It might make sense to extend the Change entry to point out that the Windows support requires VS2013 to have the information available in a slightly discoverable location.

In the fma case, I would move the new paragraph before the performance note

>  using either hardware instructions (providing full IEEE compliance) or a software
   emulation. 
   On 64-bit Cygwin, 64-bit mingw-w64 and MSVC 2017 and earlier, this function
   may be emulated owing to known bugs on limitations on these platforms.
   Note: since software emulation of the fma is costly, make sure that you are using hardware fma support if performance
   matters
   
For `round`, it is probably better to extend the `C runtime library(CRT)` abbreviation.
Otherwise, this seems sufficient to avoid potentially lengthy bug hunts from user.

dra27 committed:
Update docs for the two auto-emulated functions

Octachron commented on stdlib/float.mli:
```diff
@@ -70,9 +70,14 @@ external fma : float -> float -> float -> float =
+   emulation.
+
+   On 64-bit Cygwin, 64-bit mingw-w64 and MSVC 2017 and earlier, this function
+   may be emulated owing to known bugs on limitations on these platforms.
```
on limitations ⇒ or limitations

Octachron commented on stdlib/float.mli:
```diff
@@ -267,6 +293,50 @@ external tanh : float -> float = "caml_tanh_float" "tanh"
+
+external erfc : float -> float = "caml_erfc_float" "caml_erfc"
+  [@@unboxed] [@@noalloc]
+(** Complementary error function ([erfc x = 1.0 -. erf x]).
```
Small suggestion: `[erfc x = 1 - erf x]` since the above equation is not true with floats (which is the whole point of `erfc`).

mmottl replied:
Thanks, this suggestion seems more accurate.  I have pushed a fix.

Octachron commented:
The documentation needs to be synced with the labeled version.

mmottl committed:
Improved Float.erfc documentation

mmottl committed:
Merged float.mli changes into template

mmottl commented:
@Octachron I've also merged the float module changes into its template.

Octachron reviewed (approved):
I am grumpily ok with the current compromise: the unreliable fallback cannot be enabled without an explicit configuration flag on platforms with good C99 support. Conversely, on Windows, they might only appear on the odd combination of a recent version of OCaml and old (more than eight years old) version of VS studio.

I am thus approving this PR.

mmottl commented:
@Octachron Thanks, will there be any need for squashing commits?

Octachron commented:
I am planning to squash every commits into one (using the github interface, so no intervention needed on your part) since:
- there is some failing tests in the middle of the serie of commits
- the first commit already contains configuration changes
- the set of changes feels quite atomic

If you have any objections, don't hesitate to voice them.

mmottl commented:
Thanks, the commits are all yours to squash!

damiendoligez commented:
@Octachron Just in case you haven't noticed, this broke Inria's CI check for bootstrappability. The reason is rather tangential (pun intended) and the fix should be easy (update the `remove-sinh-primitive.patch` file).

Octachron commented:
The bootstrap CI issue is now fixed  (see da28d0e).

hhugo commented:
I just noticed that `asinh(0)` returns `-0` on windows and `0` on linux. Just saying it here in case someone cares about it.
I had to fix some test in jsoo  https://github.com/ocsigen/js_of_ocaml/pull/1135/commits/4fa155b05aa04c9faf027b23fadfdad30c9c52eb

shindere commented:
hhugo (2021/09/27 03:35 -0700):
> I just noticed that `asinh(0)` returns `-0` on windows and `0` on linux. Just saying it here in case someone cares about it.
> I had to fix some test in jsoo
> https://github.com/ocsigen/js_of_ocaml/pull/1135/commits/4fa155b05aa04c9faf027b23fadfdad30c9c52eb

@dra27, @xavierleroy, am I correct that this shouldn't happen and that
it is actually a bug which should be fixed? (I'm not proposing to fix it
myself, just asking). It feels odd to me that the result of a
computation, even on floats, depends on the OS, but I know nothing aobut
floating point arithmetics.

pmetzger commented:
Side note: there's an absolutely wonderful section of "Common Lisp: The Language" by Guy Steele where he goes through precisely this issue (that is, signs of inverse hyperbolic functions, etc.) including some nice diagrams. He also goes over branch cuts for complex valued hyperbolic functions. Worth reading if you're into this sort of thing, though I believe that diagrams of the sort are now available elsewhere.

And yes, @shindere is quite right, seeing a differing sign, even at zero, for the inverse hyperbolic sin is a sign something suboptimal _may_ be going on.

Octachron commented:
In this case, we are delegating to the C99 implementation,  I wouldn't call an OCaml bug the fact that the Windows and Linux implementation of `asinh` return different binary representations for the same number.

mmottl commented:
It definitely seems peculiar that the Windows C99 implementation would behave that way, but I agree that this should not be a concern for OCaml.  Developers using other languages on Windows platforms would probably prefer seeing consistent results when trying out such functions in OCaml, even if these results go against good numerical taste.

pmetzger commented:
In the long run, if one wants people to be able to do good numerical work in OCaml, it will be necessary to make sure it produces good, consistent results across platforms that match the best practices from the numerical analysis community. I agree that in the short run this is the problem of the C library.

shindere commented:
Perry E. Metzger (2021/09/27 09:10 -0700):
> In the long run, if one wants people to be able to do good numerical
> work in OCaml, it will be necessary to make sure it produces good,
> consistent results across platforms that match the best practices from
> the numerical analysis community. I agree that in the short run this
> is the problem of the C library.

I agree. It oculd be argued that those who want accurate results could
start by a library-implementation and that, once this gets widely
adopted it could be considered for integration to the compiler but I am
not sure how comfortable I feel with this idea.

mmottl commented:
The only reliable way to establish cross-platform consistency would be to provide our own implementation everywhere.  I don't think this is necessary or even preferable.  Implementations do not just differ in space (platform) but also over time.  Sometimes there may not be a clear winner, with some implementations approximating better in some regions than others.  Or some may sacrifice a little precision for significant speedups.  As a matter of fact, it is extremely difficult to come up with fast implementations with provable guarantees.  Distasteful as the Windows behavior for `asinh(0)` may seem, it is likely not a problem for typical applications.

Here is a short discussion concerning precision requirements in the C standard: https://stackoverflow.com/questions/20945815/math-precision-requirements-of-c-and-c-standard

You may also want to check out CRlibm and the newer [MetaLibm](http://www.metalibm.org) to see what we would be getting into if we were striving for perfection.  Btw., @Chris00 has kindly provided bindings for CRlibm: [OCaml-CRlibm](https://github.com/Chris00/ocaml-crlibm)
